Allow preview of note edits

mikehardy opened the pull request:
## Pull Request template

## Purpose / Description
The Note Editor did not allow preview, a heavily requested feature

## Fixes
Fixes #5644 / Obsoletes #5705 
Fixes #3819

## Approach
I re-use the CardTemplatePreviewer which is good for ephemeral data, with extensions to it's functionality where it can take a bundle from NoteEditor and use that to populate the card it shows with the unsaved note editor contents

## How Has This Been Tested?
API28 emulator, editing a note and hitting the new preview button to see that my edits are now shown in a preview screen. When you cancel the note edit session the edits go away. Seems to work

CardTemplateEditor preview still works as well on same API28 emulator

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/test/java/com/ichi2/anki/NonPagingCardTemplateEditor.java  1
- AnkiDroid/src/main/java/com/ichi2/anki/TemporaryModel.java  8
- AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java  9
- AnkiDroid/src/test/java/com/ichi2/anki/TestCardTemplatePreviewer.java  1
- AnkiDroid/src/test/java/com/ichi2/anki/TemporaryModelTest.java  4
- AnkiDroid/src/test/java/com/ichi2/anki/CardTemplatePreviewerTest.java  2
- AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java  9
         

Clones added
============
- AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5398275&bid=17694735)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/test/java/com/ichi2/anki/TemporaryModelTest.java  4
- AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java  2
- AnkiDroid/src/test/java/com/ichi2/anki/CardTemplatePreviewerTest.java  2
- AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java  1
- AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java  9
- AnkiDroid/src/test/java/com/ichi2/anki/TestCardTemplatePreviewer.java  1
- AnkiDroid/src/test/java/com/ichi2/anki/NonPagingCardTemplateEditor.java  1
- AnkiDroid/src/main/java/com/ichi2/anki/TemporaryModel.java  8
         

Clones added
============
- AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5398275&bid=17694735)

mikehardy commented:
Just updated to handle the case where NoteEditor has unsaved changes, but instead of Preview you go through CardTemplateEditor then Preview. Now the card, complete with it's unsaved changes will be visible there.

mikehardy commented:
Rebased now that 5151 is in. This should be ready

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -132,6 +135,30 @@ protected void onCollectionLoaded(Collection col) {
+
+            // Clear out old tags, add current tags as that can affect render
+            for (String tag : mCurrentCard.note().getTags()) {
+                mCurrentCard.note().delTag(tag);
```
Could you extract `mCurrentCard.note()`, it's fine from a performance standpoint, but makes me double-take.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -132,6 +135,30 @@ protected void onCollectionLoaded(Collection col) {
+            mCurrentCard.setDid(mNoteEditorBundle.getLong("did"));
+
+            // Clear out old tags, add current tags as that can affect render
+            for (String tag : mCurrentCard.note().getTags()) {
```
This feels like it'll cause a `ConcurrentModificationException`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -132,6 +135,30 @@ protected void onCollectionLoaded(Collection col) {
+            ArrayList<String> tags = mNoteEditorBundle.getStringArrayList("tags");
+            if (tags != null) {
+                for (String tag : tags) {
+                    if (!mCurrentCard.note().hasTag(tag)) {
```
Optional: Could extract to a method on note

mikehardy replied:
Note is libanki, I won't alter it for optional

mikehardy commented:
The core idea is the same but rather than just do it really fast rebase this time I took the time de-lint etc. All comments addressed, plus fixed a couple related / pre-existing hand-off issues between the activities, and added test cover

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:
```diff
@@ -1351,7 +1373,7 @@ private void initFieldEditText(FieldEditText editText, final int index, String[]
                 mFieldEdited = true;
                 if (index == 0) {
-                    duplicateCheck();
+                    setDuplicateFieldStyles();
```
the old name seems better (but it's not good).

mikehardy replied:
The old name had no indication it had side effects. Not it's a side-effecty verb. Changed on purpose, leaving

mikehardy commented:
> Comment; Unsure if we want to handle this, but it handles newlines differently than the reviewer.

That would be a separate defect because all 3 in the "family" now (Previewer, CardTemplatePreviewer, Reviewer) would not to sing in harmony here

mikehardy commented:
No change. Go or not?

mikehardy commented:
@Arthur-Milchior self-review in general is an anti-pattern and David marked this for 2nd approval - can you take a look? If it looks good to you clear off the status labels and hit the rebase-and-merge button otherwise I'll address whatever comments. Thanks!

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:
```diff
@@ -176,7 +176,7 @@
     private Note mEditorNote;
     public static Card mCurrentEditedCard;
-    private List<String> mSelectedTags;
+    private ArrayList<String> mSelectedTags;
```
Is there any particular reason to indicate which kinds of list is used everywhere ? Are we using any feature that is not available to standard list ?
I'd assume that it would be better to leave types as generic as possible, in order to have the freedom to change implementation as easily as possible if we ever want it.

mikehardy replied:
Yes. Per the comment on the commit. The Bundle API can put/get ArrayList types but not generic List.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -135,6 +138,32 @@ protected void onCollectionLoaded(Collection col) {
+            }
+            ArrayList<String> tagsList = mNoteEditorBundle.getStringArrayList("tags");
+            if (tagsList != null) {
+                for (String tag : tagsList.toArray(new String[0])) {
```
Any reason to transform it into an array before iterating ?
I would expect it to be easier to iterate on the list directly. In particular it would avoid to create an array just to discard it after

Arthur-Milchior replied:
I may assume that you fear the list gets modified during execution, leading to an error, maybe a comment would make it more clear

mikehardy replied:
That is precisely it, David asked for a local copy to avoid ConcurrentModificationException. I'll add a comment.

mikehardy replied:
comment clarified

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplatePreviewer.java:
```diff
@@ -135,6 +138,32 @@ protected void onCollectionLoaded(Collection col) {
+            ArrayList<String> tagsList = mNoteEditorBundle.getStringArrayList("tags");
+            if (tagsList != null) {
+                for (String tag : tagsList.toArray(new String[0])) {
+                    if (!currentNote.hasTag(tag)) {
```
Usually, you add a tag without checking whether it is already in the list of tag, so this is quite strange to read.
I don't expect it to be a problem in any way, since there is not enough tag for the inefficiency to matter. I still would feel like that it would be clearer to put in Note.java a method `setTags(List<String>)`; and leave this background task to Note.

mikehardy replied:
I am not altering Note.java (which is in libanki, and I don't want to touch it) on purpose. The tag reset is done here because if I want to make certain the tags sent to the webview (which I think are visible and thus may alter render) are exactly the tags that might be edited-but-not-saved in the NoteEditor

Arthur-Milchior replied:
It did occur multiple time that a small code was added to a libanki method, at the end of the file, indicating this is a method for us.
Removing all tags one by one is usulessly slow, even if the number of tags made it not a real problem.
At least, I believe it would make sens to put this in a function in util, or someplace like that, so that it is clear that the whole part of the code does a single thing: replace the list of tags, and if it's ever needed again, it can be reused easily

mikehardy replied:
I'm not doing that in this PR.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:
```diff
@@ -1136,9 +1136,13 @@ else if (fieldEditText.getText() != null) {
                     mReloadRequired = true;
                     mEditorNote.reloadModel();
                     if (!mEditorNote.cards().contains(mCurrentEditedCard)) {
```
I had trouble thinking how it could be true. mCurrentEditedCard was a card already existing, and card usually don't disappear unless you check for empty card. You are not supposed to be able to do it while you're reviewing.

I then realized that it can be true if you change a note type. If I'm right, I believe a comment here would be nice to explain it. If I'm wrong, it shows explanation may help here.

As a side note, I don't really get the point. Even if the card did disappear, editing the note still makes sens. Either there are other cards and we may want to edit them. Or they are no other card, and in this case the fields won't be recoverable, and it's even more important to give a strong warning to user that they'll lose data.

Actually, I had to create an add-on for anki to warn when note would be deleted because it would generate no more card. It occured multiple time to me and was always an accident, a template I did badly edit. In all cases, I would have wanted anki to warn me instead of just deleting the note

mikehardy replied:
I don't believe this functionality changed. Are you making a comment on my PR or is this a separate issue that should be logged as a separate defect?

mikehardy replied:
Also worth noting that in #5151 there was much work done to make sure that deleting a card would orphan a note. We don't allow that. I don't think it has anything to do with this PR. https://github.com/ankidroid/Anki-Android/pull/5151/commits/9777beec55b0b63f9236873fbd87f0fe8179f8c6

Arthur-Milchior replied:
Sorry, didn't see the message, I didn't thought about expending resolved conversation.

Indeed, this remark was related to the case you have not just introduced. I is not clear to me how it used to be possible. When you can enter this if branch.

mikehardy replied:
If I am in NoteEditor on for example 'Card 3', and I go to CardTemplateEditor and delete the Card 3 template, then save it and it returns to the NoteEditor, NoteEditor hits this case. And in general now it will just boot you out.

However, if you are in NoteEditor in "add note" mode (so there is no current card) and you go to CardTemplatePreviewer, you will come back here and also be in this case. But in that specific situation it's okay, there was no card to begin with so we don't need to leave.

Arthur-Milchior replied:
Thank you for the explanation.

mikehardy commented:
@Arthur-Milchior I've addressed all your comments to date, I don't have any changes I'm planning on making at the moment besides the clarifying comment. I'm not going to alter Notes in libanki, I'm unconcerned about unlikely tag performance in a feature that didn't exist before. If you have anything else let me know.

timrae commented:
Awesome! Users will never know how much hard work went into giving them this feature, but I know. Good job Mike! 😆

Arthur-Milchior commented:
I don't have anything else. Thank you for the comments.
The point of my comment was not about efficiency, there is no bottleneck here. It is to put a single idea "putting this exact list of tag to this note" in a single function to make code simpler to read.

mikehardy commented:
Yeah - not a bad idea @Arthur-Milchior just trying to stay focused on the PR. If I shave every yak on every PR I'll never get anything done :-) https://americanexpress.io/yak-shaving/

mikehardy commented:
@timrae yeah - getting 5151 out unblocked quite a few things. The pass-data-via-temp-file way to avoid Bundle data overflow still has one crash it can fix too, I forget where it is but I want to say it's fields. Might actually apply here! But first I just wanted to use the new "display an ephemeral card" capability for something users actually cared about ;-)
